src/py21cmfast/drivers/coeval.py

```diff
@@ -687,65 +678,51 @@ def _redshift_loop_generator(
                 **kw,
             )
 
-            if prev_coeval is not None:
-                with contextlib.suppress(OSError):
-                    prev_coeval.perturbed_field.purge(force=always_purge)
-
-            if this_pthalo is not None:
-                with contextlib.suppress(OSError):
-                    this_pthalo.purge(force=always_purge)
-
-            # we only need the SFR fields at previous redshifts for XraySourceBox
-            if this_halobox is not None:
-                with contextlib.suppress(OSError):
-                    this_halobox.prepare(
-                        keep=[
-                            "halo_sfr",
-                            "halo_sfr_mini",
-                            "halo_xray",
-                            "log10_Mcrit_MCG_ave",
-                        ],
-                        force=always_purge,
-                    )
-
-            logger.debug(f"PID={os.getpid()} doing brightness temp for z={z}")
-
-            _bt = sf.brightness_temperature(
+            this_bt = sf.brightness_temperature(
                 ionized_box=this_ionized_box,
                 perturbed_field=this_perturbed_field,
                 spin_temp=this_spin_temp,
                 write=write.brightness_temp,
                 **iokw,
             )
 
-            if inputs.astro_options.PHOTON_CONS_TYPE == "z-photoncons":
-                # Updated info at each z.
-                photon_nonconservation_data = _get_photon_nonconservation_data()
-
             this_coeval = Coeval(
                 initial_conditions=initial_conditions,
                 perturbed_field=this_perturbed_field,
                 ionized_box=this_ionized_box,
-                brightness_temperature=_bt,
+                brightness_temperature=this_bt,
                 ts_box=this_spin_temp,
                 halobox=this_halobox,
                 photon_nonconservation_data=photon_nonconservation_data,
             )
 
+            # yield before the cleanup, so we can get at the fields before they are purged
+            yield iz, this_coeval
+
+            # We purge previous fields and those we no longer need
+            if prev_coeval is not None:
+                prev_coeval.perturbed_field.purge(force=True)
+                if inputs.matter_options.USE_HALO_FIELD:
+                    prev_coeval.halobox.prepare_for_next_snapshot(force=True)
```

Comment:
Weren't we anyway purging the prev_coeval simply because it goes out of scope?

Reply:
The references are still stored in the `perturbed_field` list, which is loaded at the start of the loop